### PR TITLE
Install additional-compatible-tkr config map as part of tkr-source-controller package installation

### LINKED ISSUE #2597: Install additional-compatible-tkr related config map when creating/upgrading management-cluster
**Describe the feature request**

In conjunction to #2592, when creating management-cluster with `package-based-lcm` feature-flag enabled, we need to create a config map in the bootstrap as well and management-cluster to mark a specific TKR compatible that will be used for management-cluster creation and upgrade.

For this, we can create this ConfigMap as part of TKG composite package deployment and update the configMap during the management-package upgrade when upgrading the management-cluster to newer version.
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: tkr-metadata
  namespace: tkg-system
  labels:
    run.tanzu.vmware.com/additional-compatible-tkrs: ""
data:
  tkrVersions: |
    - v1.23.6+vmware.1-tkg.1
```

The version of the TKR to specify here will be determined from the default BoM file and will be passed as a dataValue to the TKG composite package.

**Describe alternatives you've considered**

**Affected product area (please put an X in all that apply)**

- [ ] APIs
- [ ] Addons
- [x] CLI
- [ ] Docs
- [ ] IAM
- [ ] Installation
- [ ] Plugin
- [ ] Security
- [ ] Test and Release
- [ ] User Experience

**Additional context**

### PR DESCRIPTION
**Pull request by anujc25**
### What this PR does / why we need it
- Install additional-compatible-tkr config map as part of tkr-source-controller package installation

- This change will create this ConfigMap as part of TKG composite package deployment and update the configMap during the management-package upgrade when upgrading the management-cluster to newer version.
- So, that tkr-source-controller will always have one compatible tkr available while creating/upgrading management cluster

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: tkr-metadata
  namespace: tkg-system
  labels:
    run.tanzu.vmware.com/additional-compatible-tkrs: ""
data:
  tkrVersions: |
    - v1.23.6+vmware.1-tkg.1
```

- This PR also updates the hardcoded tkr-repository configuration that we are passing to the tkr-source-controller package to use data from BoM file

### Which issue(s) this PR fixes

Fixes #2597, Related to #2592

### Describe testing done for PR
- Removed the `_ADDITIONAL_MANAGEMENT_COMPONENT_CONFIGURATION_FILE` from the pipeline and ran the AWS clusterclass e2e test
- Seems like the pipeline is successful and it verifies this change

### Release note

```release-note
Install additional-compatible-tkr config map as part of tkr-source-controller package installation
```

### PR Checklist

- [x] Squash the commits into one or a small number of logical commits
      
- [x] Use good commit [messages](https://github.com/vmware-tanzu/tanzu-framework/blob/main/CONTRIBUTING.md)
- [x] Ensure PR contains terms all contributors can understand and links all contributors can access

### CODE REVIEW

File: packages/tkr-source-controller/bundle/config/upstream/default-compatible-tkr.yaml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+---
+apiVersion: v1
+data:
+  tkrVersions: #@ yaml.encode(getCompatibleTKRs())
+kind: ConfigMap
+metadata:
+  labels:
+    run.tanzu.vmware.com/additional-compatible-tkrs: ""
+  name: tkg-compatibility-versions
+  namespace: tkg-system
```
Comment (prkalle) [inline packages/tkr-source-controller/bundle/config/upstream/default-compatible-tkr.yaml:17]: namespace value should be fetched from package value (`#@ data.values.namespace`)

**Review by prkalle — COMMENTED** [review]
LGTM. Thanks for addressing the hack fix!
I have one minor comment.

File: pkg/v1/tkg/managementcomponents/helper.go [hunk 2]
```diff
@@ -21,18 +23,18 @@ const (
 )
 
 // GetTKGPackageConfigValuesFileFromUserConfig returns values file from user configuration
-func GetTKGPackageConfigValuesFileFromUserConfig(managementPackageVersion string, userProviderConfigValues map[string]interface{}) (string, error) {
+func GetTKGPackageConfigValuesFileFromUserConfig(managementPackageVersion string, userProviderConfigValues map[string]interface{}, tkgBomConfig *tkgconfigbom.BOMConfiguration) (string, error) {
```
Comment (saji-pivotal) [inline pkg/v1/tkg/managementcomponents/helper.go:26]: We're adding an additional input to this function. Are there corresponding unit tests that need to get updated / added?

**Review by saji-pivotal — COMMENTED** [review]
Thanks for the changes @anujc25  I do have a question on adding unit tests

**Comment by vuil** [comment]
Please also update : docs/cli/tkgctl/management-cluster-using-clusterclass.md 
to mention this config map. If it helps feel free to reference issue #2592 as well.

File: pkg/v1/tkg/managementcomponents/helper.go [hunk 3]
```diff
@@ -21,18 +23,18 @@ const (
...
 	// TODO: Temporary hack(hard coded values) to configure TKR source controller package values. This should be replaced with the logic
 	// that fetches these values from tkg-bom(for bom related urls) and set the TKR source controller package values
 	var tkrRepoImagePath string
 	providerType := userProviderConfigValues[constants.ConfigVariableProviderType]
 	switch providerType {
 	case constants.InfrastructureProviderVSphere:
-		tkrRepoImagePath = "projects-stg.registry.vmware.com/tkg/tkr-repository-vsphere-nonparavirt"
+		tkrRepoImagePath = fmt.Sprintf("%s/%s", tkgBomConfig.ImageConfig.ImageRepository, tkgBomConfig.TKRPackageRepo.VSphereNonparavirt)
```
Comment (vuil) [inline pkg/v1/tkg/managementcomponents/helper.go:33]: I think this is an additional fix we did to leverage the bom data for source controller parameters. Maybe add a note of that in the PR description, commit message.

File: pkg/v1/tkg/managementcomponents/helper.go (hunk 2 above)
Comment (anujc25) [reply pkg/v1/tkg/managementcomponents/helper.go:27]: Thanks for catching this. I added unit for this function explicitly.

File: docs/cli/tkgctl/management-cluster-using-clusterclass.md [hunk 4]
```diff
@@ -6,7 +6,6 @@ Below is the sequence of management-cluster creation using clusterclass.
 1. Deploy `kapp-controller` on bootstrap cluster
 1. Deploy `providers` on bootstrap cluster
 1. Deploy `TKG meta package` to bootstrap cluster
```
Comment (vuil) [inline docs/cli/tkgctl/management-cluster-using-clusterclass.md:8]: nit: name change?

Also in my earlier comment I was referring to mentioning the tkr compat config map somewhere in this sequence to call out this complication with the MC cluster bootstrap.

We can address in a followup as well.

**Review by vuil — APPROVED** [review]
lgtm, thanks!
Still have a couple of doc nits, but we can address in the followup before merging to main.

File: docs/cli/tkgctl/management-cluster-using-clusterclass.md (hunk 4 above)
Comment (anujc25) [reply docs/cli/tkgctl/management-cluster-using-clusterclass.md:8]: Updated

**Review by saji-pivotal — APPROVED** [review]

**Pull request merged**